presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java

```diff
@@ -290,24 +294,27 @@ public synchronized void scheduleRequestIfNecessary()
             return;
         }
 
-        long neededBytes = bufferCapacity - bufferBytes;
-        if (neededBytes <= 0) {
-            return;
-        }
-
-        int clientCount = (int) ((1.0 * neededBytes / averageBytesPerRequest) * concurrentRequestMultiplier);
-        clientCount = Math.max(clientCount, 1);
-
-        int pendingClients = allClients.size() - queuedClients.size() - completedClients.size();
-        clientCount -= pendingClients;
-
-        for (int i = 0; i < clientCount; i++) {
+        long availableBytes = bufferCapacity - bufferBytes;
+        while (reservedBytes.get() < availableBytes) {
             HttpPageBufferClient client = queuedClients.poll();
             if (client == null) {
                 // no more clients available
                 return;
             }
-            client.scheduleRequest();
+            // reservedBytes cannot be greater than availableBytes as this is the only place to bump up reservedBytes
+            long maxBytes = Math.min(client.getTotalPendingSize().toBytes(), availableBytes - reservedBytes.get());
+            reservedBytes.addAndGet(maxBytes);
+            client.scheduleRequest(new DataSize(maxBytes, BYTE));
+        }
+
+        // drain all the clients that do not have a pending page
```

Comment:
I don't understand why we are doing this.  Can you comment on the purpose of this section.